core/src/main/scala/org/ensime/core/javac/JavaCompiler.scala

```diff
@@ -211,5 +218,23 @@ private class NotSoSimpleJavaFileObject(
     simpleName: String,
     kind: JavaFileObject.Kind
   ): Boolean = false // this can probably be optimised, but the intended javac use is not clear
-
+  override def equals(other: Any): Boolean = this.f match {
+    case RawFile(rf) => other match {
+      case p: Path => rf.toString == p.toString
+      case r: RawFile => rf.toString == r.file.toString
+      case _ => false
+    }
+    case ArchiveFile(p, e) => other match {
+      case af: ArchiveFile => af.entry.toString == e && af.jar.toFile.toString == p.toFile.toString
```

Comment:
all this `.toString` magic is cryptic to me, can we please break this into functions with unit tests that assert the behaviour?